src/Service.js

```diff
@@ -376,26 +376,33 @@ Service_.prototype.handleCallback = function(callbackRequest) {
  *     otherwise.
  */
 Service_.prototype.hasAccess = function() {
-  return this.lockable_(function() {
-    var token = this.getToken();
-    if (!token || this.isExpired_(token)) {
-      try {
-        if (token && this.canRefresh_(token)) {
-          this.refresh();
-        } else if (this.privateKey_) {
-          this.exchangeJwt_();
-        } else if (this.grantType_) {
-          this.exchangeGrant_();
-        } else {
+  var token = this.getToken();
+  if (!token || this.isExpired_(token)) {
+    return this.lockable_(function() {
+      // Get the token again, bypassing the local memory cache.
+      token = this.getToken(true);
+      // Check to see if the token is still missing or expired, as another
+      // execution may have refreshed it while we were waiting for the lock.
+      if (!token || this.isExpired_(token)) {
```

Comment:
- We can remove 1 layer of nesting.
- I usually don't add a space between `catch` and `(` but can't find a rule either way.

Here's rough pseudocode with comments for you:
```js
if (token && !this.isExpired_(token)) return true; // token is good
// ...otherwise
try { // try to refresh
  if ...
  else if ...
  else {
    return false;
  }
} catch(e) {
  ...
}
```

Reply:
Done